tensorflow/python/ops/embedding_ops.py

```diff
@@ -495,31 +501,47 @@ def embedding_lookup_sparse(params,
     params = list(params)  # Iterate to get the underlying Variables.
   if not isinstance(params, list):
     params = [params]
-  if not isinstance(sp_ids, sparse_tensor.SparseTensor):
-    raise TypeError(f"sp_ids must be SparseTensor, got {type(sp_ids)}")
+  if not isinstance(sp_ids, sparse_tensor.SparseTensor) \
+            and not isinstance(sp_ids, ragged_tensor.RaggedTensor):
+    raise TypeError(f"sp_ids must be SparseTensor or RaggedTensor, \
+            got {type(sp_ids)}")
   ignore_weights = sp_weights is None
   if not ignore_weights:
-    if not isinstance(sp_weights, sparse_tensor.SparseTensor):
-      raise TypeError(f"sp_weights must be either None or SparseTensor,"
-                      f"got {type(sp_weights)}")
-    sp_ids.values.get_shape().assert_is_compatible_with(
-        sp_weights.values.get_shape())
-    sp_ids.indices.get_shape().assert_is_compatible_with(
-        sp_weights.indices.get_shape())
-    sp_ids.dense_shape.get_shape().assert_is_compatible_with(
-        sp_weights.dense_shape.get_shape())
+    if type(sp_weights) is not type(sp_ids):
+      raise TypeError(f"sp_ids must be of the same type as sp_weights.")
+    if isinstance(sp_weights, sparse_tensor.SparseTensor):
+      sp_ids.values.get_shape().assert_is_compatible_with(
+          sp_weights.values.get_shape())
+      sp_ids.indices.get_shape().assert_is_compatible_with(
+          sp_weights.indices.get_shape())
+      sp_ids.dense_shape.get_shape().assert_is_compatible_with(
+          sp_weights.dense_shape.get_shape())
+    else:
+      sp_ids.values.get_shape().assert_is_compatible_with(
+          sp_weights.values.get_shape())
+      sp_ids.get_shape().assert_is_compatible_with(
+          sp_weights.get_shape())
     # TODO(yleon): Add enhanced node assertions to verify that sp_ids and
     # sp_weights have equal indices and shapes.
 
   with ops.name_scope(name, "embedding_lookup_sparse",
                       params + [sp_ids]) as name:
-    segment_ids = sp_ids.indices[:, 0]
 
-    ids = sp_ids.values
-    ids, idx = array_ops.unique(ids)
+    if isinstance(sp_ids, sparse_tensor.SparseTensor):
+      segment_ids = sp_ids.indices[:, 0]
```

Comment:
In your case, is the `sp_ids.indices` a huge tensor and `sp_ids.indices[:, 0]` is super slow?

Reply:
Yes, that’s a major performance limitation.